### ISSUE TITLE
upgrade to babeld 1.8.1 (or later version)

### ISSUE THREAD

**Issue by jhpoelen** [sudomesh/bugs#33]
Currently, we are using a custom fork of babeld. I believe that the reason for forking was to add the ability to  add / remove interfaces at runtime (e.g., options ```-a``` and ```-x```), without having to restart babeld. Another reason was to have the ability to dump babeld state. Many modifications were made and the functionality was not merged back into mainline babeld.

In context of a segmentation fault bug (https://github.com/sudomesh/bugs/issues/31), @jech, the author of babeld suggested to upgrade to a recent version main-line (see https://github.com/sudomesh/bugs/issues/31#issuecomment-388341089). Apparently, babeld version after 1.7.0 have the ability to add/remove interfaces without having to restart babeld. Also, it appears that some debugging ("dump") command can be used to extract state info from babeld process. Understandable, the author is willing to help debug the issue only after we start using a mainline version of babeld.

Suggest to switch to version 1.8.1 of babeld and deprecate our own babeld fork. Note that our current fork was last synced on 31 March 2015, just before the release of babeld 1.6.0 .

**Comment by jech** [comment]
> In context of a segmentation fault bug (#31), @jech, the author of babeld
> suggested to upgrade to a recent version main-line (see #31 (comment)).

Please.

> the author is willing to help debug the issue only after we start using
> a mainline version of babeld.

I'd be quite willing to help in either case, but I'd be grateful if you
could upgrade first.

**Comment by paidforby** [comment]
Since there are a lot of interdependent changes that need to happen to do this upgrade, I suggest we coordinate some the work.
- [ ] sudowrt-firmware needs to be re-built with main fork of babeld, https://github.com/sudomesh/sudowrt-firmware/blob/master/openwrt_config/packages#L45
- [ ] Any sudomesh/babeld dependent configs/scripts in firmware need to be removed or refactored, potential problem files, https://github.com/sudomesh/sudowrt-firmware/blob/zeroconf/files/opt/notdhcpserver/hook.sh#L48 and https://github.com/sudomesh/sudowrt-firmware/blob/zeroconf/files/opt/mesh/templates/etc/init.d/meshrouting
Note: I mostly interested in this in the context of zeroconf, of course, necessary changes will also need to be made to makenode
- [ ] exitnode needs to clone main babeld (or just apt-get it?), https://github.com/sudomesh/exitnode/blob/master/create_exitnode.sh#L77
- [ ] sudomesh/babeld dependent scripts in exitnode need to be refactored https://github.com/sudomesh/exitnode/blob/master/create_exitnode.sh#L113

I guess the first step would be to start a new build of the firmware with main babeld. Then spin up a test exit node also with main babeld. Finally, get them to babel by testing changes manually.

**Comment by jhpoelen** [comment]
A proposal is to split into two babeld upgrade issues: one for exit node, one for firmware.

**Comment by jech** [comment]
> Here, I am assuming that the babeld protocol has not changed.

The core protocol remains compatible.

The source-specific extension will change in an incompatible manner in
version 1.9.  If you don't know what source-specific routing is, then
you're not concerned.

> Also, I don't see a reason for urgency in upgrading the firmware version
> of babeld, because we have not observed any crashes on the routers.

The core protocol has remained compatible since version 0.90, back in
2008, so you should be fine.  The full changelog is here:

  https://github.com/jech/babeld/blob/master/CHANGES

**Comment by bennlich** [comment]
Planned out an upgrade with @eenblam and @Juul today towards the end of the hacking workgroup. I just implemented the upgrade on the HE exit node, and everything seems to be working! My home node succeeded in getting routes.

See https://github.com/sudomesh/exitnode/pull/15 for changes.

**Comment by bennlich** [comment]
@jech I noticed during the upgrade that `babeld` wants to be started with at least one interface. My current solution is to start it with a fake interface name: `babeld dummyinterface`. This seems to work, but I wonder if I'm missing a better way.

**Comment by paidforby** [comment]
@jech is there any harm in running `babeld lo`? Then there isn't this weird "dummyinterface" idea. Perhaps you could suggest that as a solution to people would like to start/restart babeld arbitrarily, yet find it confusing to run it on an imaginary interface (then again, isn't loopback a kinda imaginary interface?)

**Comment by bennlich** [comment]
> I could either make the error into a warning, or add a command-line option that allows running with no interfaces.

@jech Either of these sounds good to me. Am sympathetic with your reasoning re: people being confused.

@paidforby I prefer `dummyinterface` to `lo`. `lo` seems like it could actually be meaningful, but isn't, which I think is confusing (e.g. sometimes you really do want to bind things to the loopback interface). But either of @jech's suggestions seems 
much better to me than passing in an interface you don't care about.

**Comment by bennlich** [comment]
@jech after the upgrade, it looks like we're still seeing the same memory leak :-( (though maybe less frequently). from the logs:

```
Sep  7 10:59:13 exit0 babeld[3975]: Warning: cannot save old configuration for l2tp1056-1.
Sep  7 10:59:13 exit0 babeld[3975]: setsockopt(IPV6_JOIN_GROUP): Cannot allocate memory
Sep  7 10:59:13 exit0 babeld[3975]: setsockopt(IPV6_LEAVE_GROUP): Cannot assign requested address
Sep  7 10:59:13 exit0 babeld[3975]: Warning: cannot restore old configuration for l2tp1056-1.
Sep  7 10:59:40 exit0 babeld[3975]: Warning: cannot save old configuration for l2tp1056-1.
```

**Comment by bennlich** [comment]
Closing this issue, as we've been running babeld 1.8.2 on our exitnodes for a while now.

Current thinking about the memory leak mentioned [above](https://github.com/sudomesh/bugs/issues/33#issuecomment-419612326) is that it has to do with the tunnel interface disappearing before babeld has time to clean up some of the sockopts that get allocated. I think the fix for this is to add support to tunneldigger for a blocking tunnel-down hook, where babeld can clean up the resources it was using before the interface disappears. This should go in a new issue.
